x-pack/plugins/lens/public/xy_visualization_plugin/xy_suggestions.test.ts

```diff
@@ -0,0 +1,120 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { getSuggestions } from './xy_suggestions';
+import { TableColumn } from '../types';
+
+describe('xy_suggestions', () => {
+  function numCol(columnId: string): TableColumn {
+    return {
+      columnId,
+      operation: {
+        dataType: 'number',
+        id: `avg_${columnId}`,
+        label: `Avg ${columnId}`,
+        isBucketed: false,
+      },
+    };
+  }
+
+  function strCol(columnId: string): TableColumn {
+    return {
+      columnId,
+      operation: {
+        dataType: 'string',
+        id: `terms_${columnId}`,
+        label: `Top 5 ${columnId}`,
+        isBucketed: true,
+      },
+    };
+  }
+
+  function dateCol(columnId: string): TableColumn {
+    return {
+      columnId,
+      operation: {
+        dataType: 'date',
+        id: `date_histogram_${columnId}`,
+        isBucketed: true,
+        label: `${columnId} histogram`,
+      },
+    };
+  }
+
+  test('ignores invalid combinations', () => {
+    expect(
+      getSuggestions({
+        tableColumns: {
+          1: [dateCol('a')],
+          2: [strCol('foo'), strCol('bar')],
+        },
+      })
+    ).toEqual([]);
+  });
+
+  test('suggests a basic x y chart with date on x', () => {
+    expect(
+      getSuggestions({
+        tableColumns: {
+          0: [numCol('a')],
+          1: [numCol('bytes'), dateCol('date')],
+        },
+      })
+    ).toMatchSnapshot();
+  });
+
+  test('suggests a split x y chart with date on x', () => {
+    expect(
+      getSuggestions({
+        tableColumns: {
+          1: [numCol('price'), numCol('quantity'), dateCol('date'), strCol('product')],
+        },
+      })
+    ).toMatchSnapshot();
+  });
+
+  test('supports multiple suggestions', () => {
+    expect(
+      getSuggestions({
+        tableColumns: {
+          1: [numCol('price'), dateCol('date')],
+          2: [numCol('count'), strCol('country')],
+        },
+      })
+    ).toMatchSnapshot();
+  });
+
+  test('handles two numeric values', () => {
+    expect(
+      getSuggestions({
+        tableColumns: {
+          1: [numCol('quantity'), numCol('price')],
+        },
+      })
+    ).toMatchSnapshot();
+  });
+
+  test('places non-numeric columns on x', () => {
+    expect(
+      getSuggestions({
+        tableColumns: {
+          1: [
+            numCol('num votes'),
+            {
+              columnId: 'mybool',
+              operation: {
+                dataType: 'boolean',
+                id: 'mybool',
+                isBucketed: false,
+                label: 'Yes / No',
+              },
+            },
+          ],
+        },
+      })
+    ).toMatchSnapshot();
```

Comment:
A lot of these snapshot-oriented tests are hard to read for me because I have to correlate with another file. It seems like there is a limited set of "important" fields in the state that you could extract with a test helper, and it would be really helpful for keeping these assertions meaningful. I would expect for example, that you would care about:

* x accessor
* y accessors
* split accessors
* series accessors
* seriesType

One possible approach would be to write a test helper here like `getAssertionPropsFromState(state)` and then you could do something like:

```typescript
expect(suggestions.length).toEqual(2);
const props = getAssertionPropsFromState(suggestions[0].state);
expect(props.xAccessor).toEqual('country');
expect(props.yAccessors).toEqual(['count']);
expect(props.seriesType).toEqual('bar');
```

Reply:
I both agree and disagree. I think the snapshot tests are really handy and convenient, and allow us to easily see when anything changes in the output, thus catching unintentional changes. Intentional changes that cause breaks are trivial to adjust (type "u" in the terminal, and Jest updates the snapshots), and then you can diff to make sure the new snapshots look right. I think we should keep them.

I do agree with the downside of them living in an external file, though. I wish Jest would inline them as JS string literals or something similar. But I understand why it doesn't.

Reply:
Isn't there also `toMatchInlineSnapshot` for this use case? https://jestjs.io/docs/en/expect.html#tomatchinlinesnapshotpropertymatchers-inlinesnapshot